### PR TITLE
[SPARK-4134] [SPARK-7835] [WIP] Dynamic allocation: Tone down kill error messages

### PR DESCRIPTION
**Pull request by andrewor14**
**[SPARK-4134]:** We should not log `ERROR` or `WARNING` if executors are killed by the user.
**[SPARK-7835]:** Refactor `HeartbeatReceiverSuite` to increase coverage (needed by SPARK-4134)

~60% of this patch is test code.

---

Right now we get a bunch of scary error messages if the user kills executors manually (or uses dynamic allocation to do so):

screen shot 2015-05-20 at 3 42 38 pm

This patch tones down these error messages, since these are really not errors. Now the output looks like:

screen shot 2015-05-21 at 1 17 02 pm

I tested this on a real YARN cluster with #6301.

### CODE REVIEW

**Comment by harishreedharan** [comment]
Not sure the whole `Logging` trait etc will allow us to hide the logging per class, but if it does you could change the configuration to log only `ERROR` level messages for `ReliableDeliverySupervisor`

**Comment by harishreedharan** [comment]
From what I see here: https://github.com/akka/akka/blob/d9db42b75715cb2ca98c84efbdb8666cb046bfa2/akka-remote/src/main/scala/akka/remote/Endpoint.scala ignoring logging messages below `ERROR` seems ok for that class.

**Comment by andrewor14** [comment]
@harishreedharan yeah I'm just going to bump it to `ERROR` in the log4j conf. After an offline discussion with @zsxwing we decided that it's safe to do so.

**Comment by andrewor14** [comment]
By the way I updated the screenshot and fixed a heartbeat receiver race condition that was introduced in this patch previously.

**Comment by andrewor14** [comment]
It would be good for @zsxwing and @vanzin to have a look, since this modifies some code both of you have touched in the recent past.

File: core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala [hunk 1]
```diff
@@ -369,24 +369,47 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
...
+        if (executorDataMap.contains(id)) {
+          executorIdsToKill += id
+        } else {
+          logWarning(s"Executor to kill $id does not exist!")
+        }
+      }
+
+      executorsPendingToRemove ++= executorIdsToKill
+
+      numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala]: @vanzin I believe this is functionally equivalent to what you had in the old code (L384). I believe the only difference is that I moved the `askWithReply` out of the synchronized block. Please let me me know if this is not the case.
```
val newTotal = (numExistingExecutors + numPendingExecutors - executorsPendingToRemove.size
  - filteredExecutorIds.size)
...
executorsPendingToRemove ++= filteredExecutorIds
```

**Comment by vanzin** [comment]
Looks sane to me.

**Comment by andrewor14** [comment]
After testing further I found that the race condition wasn't fully resolved. This should be actually fixed in the latest commit.

File: core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala [hunk 2]
```diff
@@ -92,24 +102,38 @@ private[spark] class HeartbeatReceiver(sc: SparkContext)
...
+    case ExecutorRemoved(executorId) =>
+      executorLastSeen.remove(executorId)
     case TaskSchedulerIsSet =>
       scheduler = sc.taskScheduler
   }
 
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case ExpireDeadHosts =>
+      expireDeadHosts()
+      context.reply(true)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala]: moved this here to keep the tests deterministic

File: core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala [hunk 3]
```diff
@@ -92,24 +102,39 @@ private[spark] class HeartbeatReceiver(sc: SparkContext)
...
+      executorLastSeen.remove(executorId)
   }
 
   override def receiveAndReply(context: RpcCallContext): PartialFunction[Any, Unit] = {
+    case TaskSchedulerIsSet =>
+      scheduler = sc.taskScheduler
+      context.reply(true)
+    case ExpireDeadHosts =>
+      expireDeadHosts()
+      context.reply(true)
```
Comment (andrewor14) [inline core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala]: I moved these here to increase test determinism

File: core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala [hunk 4]
```diff
@@ -43,15 +43,25 @@ private[spark] case class Heartbeat(
...
 /**
  * Lives in the driver to receive heartbeats from executors..
  */
-private[spark] class HeartbeatReceiver(sc: SparkContext)
-  extends ThreadSafeRpcEndpoint with Logging {
+private[spark] class HeartbeatReceiver(sc: SparkContext, clock: Clock)
+  extends ThreadSafeRpcEndpoint with SparkListener with Logging {
+
+  def this(sc: SparkContext) {
+    this(sc, new SystemClock)
```
Comment (zsxwing) [inline core/src/main/scala/org/apache/spark/HeartbeatReceiver.scala]: Could you make `SystemClock` as a default value to save one constructor, please?

**Comment by andrewor14** [comment]
I discovered a few more issues with this patch. Putting this on hold for now.

File: core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala [hunk 5]
```diff
@@ -17,65 +17,145 @@
...
-import org.apache.spark.util.RpcUtils
-import org.scalatest.concurrent.Eventually._
+import org.apache.spark.executor.TaskMetrics
+import org.apache.spark.rpc.RpcEndpointRef
+import org.apache.spark.scheduler._
+import org.apache.spark.storage.BlockManagerId
+import org.apache.spark.util.ManualClock
 
-class HeartbeatReceiverSuite extends FunSuite with LocalSparkContext {
+class HeartbeatReceiverSuite
```
Comment (andrewor14) [inline core/src/test/scala/org/apache/spark/HeartbeatReceiverSuite.scala]: I'm going to factor the changes in this class out to a separate patch

**Comment by andrewor14** [comment]
I'm going to open a new patch later once I spend some time to figure out the issue with this one. In the mean time let's not block the refactoring of the `HeartbeatReceiverSuite` on this.

**Pull request closed without merging**